[MRG+1] MultiLabelBinarizer ignore unkown class in transform

rragundez opened the pull request:
#### Reference Issues/PRs
Fixes #10410 

#### What does this implement/fix? Explain your changes.
Like with other transformers I would expect that `transform` uses what is picked up/learned during `fit` (e.g. `sklearn.preprocessing.StandardScaler` uses the mean and variance learned during `fit`), to my surprise this was not the case with `MultiLabelBinarizer`.

Example:
```python
from sklearn.preprocessing import MultiLabelBinarizer
mlb = MultiLabelBinarizer()
y_train = [['a'],['a', 'b'], ['a', 'b', 'c']]
mlb.fit(y_train)
y_test = [['a'],['b', 'c'],['d']]
mlb.transform(y_test)
>>> KeyError: 'd'
```
My case was that I used a container to train the model with a pre-processing step that uses `MultiLabelBinarizer`, build pipeline and output the pipeline as pkl file. Another container picks it up and `predicts` on new data. To my surprise the pre-processing pipeline breaks if an unseen label is received. To fix this I had to write an ugly hack using `MultiLabelBinarizer`'s `classes_` attribute.

This change addresses that problem. Now `transform` only uses the seen during `fit`. If the parameter `classes` was given on initialization it will use that. If `fit_transform`  is used it respects the optimization already in place.

Example with change:
```python
from sklearn.preprocessing import MultiLabelBinarizer
mlb = MultiLabelBinarizer()
y_train = [['a'],['a', 'b'], ['a', 'b', 'c']]
mlb.fit(y_train)
y_test = [['a'],['b', 'c'],['d']]
mlb.transform(y_test)
>>> array([[1, 0, 0],
       [0, 1, 1],
       [0, 0, 0]])
```
#### Any other comments?
The Issue 10410 that this change fixes, mentions the idea of adding an argument `ignore_unseen`. In my opinion this is not the a good solution:
- It does not respect the pattern of other transformers, scalers, etc.
- It is not intuitive
- It adds unnecessary complexity

rragundez committed:
Ignore unknown classes in MultiLabelBinarizer transform

This only modifies the behaviour when using transform after using
fit or if the classes argument was used during initialization.

When using transform it can come after using fit or directly from
fit_transform. In the former case makes sense to only use the
classes learned by using fit. In the latter case all classes
are used.

rragundez committed:
Update tests for unkown class in MultiLabelBinarizer

Previously this tests will check for a KeyError. MultiLabelBinarizer
now ignores unknown classes.

jnothman reviewed (commented):
I agree that this will be useful for some tasks. I would suggest, though, that giving users an error is useful in many cases. We should inform the user is informed when their data is problematic.

I would prefer an option `handle_unknown='ignore'/'error'` (with error being the default).

rragundez commented on sklearn/preprocessing/label.py:
```diff
@@ -794,8 +795,16 @@ def _transform(self, y, class_mapping):
         indptr = array.array('i', [0])
+        empty_mapping = False if class_mapping else True
         for labels in y:
-            indices.extend(set(class_mapping[label] for label in labels))
```
Hi @jnothman thanks for your comments. 
I agree that it is sometimes useful to give the user an error, but I completely disagree this is one of those cases, as the change I'm proposing only kicks in when the user EXPLICITLY has `fitted` the object and then `transforms` or given an explicit collection of `classes` to use in the object initialization and then. 
If you want to go to the real of  "when the data is problematic" it is a lost battle as there are so many exceptions that is impossible to catch in a simple pragmatic and intuitive API.

Is the responsibility of the user to check its data, and for example use the `classes` parameters to include all possible classes. Or if it is just expanding every class on every dataset then the `fit_transform` would allow that. Could you give me an example in practice you have used or seen where the solutions I'm proposing together with `fit_transform` don't work? I haven't seen one.

Your code doesn't work. I won't go to the specifics but `class_mapping` can be a `defaultdict` with a `default_factory`, which mean it increases on demand. I looked for other more compressed forms to write the code (I specially dislike line 798)  but unfortunately I couldn't find one which was simple, easy to understand, worked for `transform` and `fit_transform`, and will require changing only in one function. But if you can think of one please be my guest :)

rragundez replied:
I think it is a good idea to raise warning on the missing class btw

jnothman replied:
> Is the responsibility of the user to check his/her data, and for example use the classes parameters to include all possible classes.

Or it is the responsibility of the user to clean his/her data to avoid unknowns. But I do see what you mean, in a random sample context.

I suppose in this context I'm okay with changing the error to a warning. Sometimes you'd be surprised what behaviour users expect to remain backwards compatible, though! Please make that change, and we'll see what other core devs think.

> Your code doesn't work. I won't go to the specifics but class_mapping can be a defaultdict with a default_factory, which means it increases on demand.

You're right. I meant `not empty_mapping` not `class_mapping`: https://github.com/rragundez/scikit-learn/compare/mlb-unkown-class...jnothman:ignore-unknown?expand=1

rragundez replied:
I agree that for backwards compatibility yours is the solution that bridges the gap, and I understand that even though we both might not agree on this being the best solution perhaps is the best solution for the community. So I'll add the change with under protest :)

I'll also add your piece of code, thanks!

jnothman committed:
Blah

rragundez committed:
Merge pull request #1 from jnothman/ignore-unknown

Refactor empty mapping logic

rragundez committed:
Add comment on using known classes by MultiLabelBinarizer

rragundez committed:
Add Waring if unkown classes are found

rragundez commented:
@jnothman I've added the warning. 

Example:
```python
from sklearn.preprocessing import MultiLabelBinarizer
mlb = MultiLabelBinarizer()

y_train = [['a'],['a', 'b'], ['a', 'b', 'c']]
mlb.fit(y_train)

y_test = [['a'],['b'],['d', 'f']]
mlb.transform(y_test)
>>> Warning: Unkown class(es) found 'f', 'd' will be ignored.
```

jnothman commented on sklearn/preprocessing/label.py:
```diff
@@ -794,7 +795,19 @@ def _transform(self, y, class_mapping):
+                labels = set(labels)  # in case labels is a generator
+                missing = labels - set(class_mapping)
+                if missing:
+                    raise Warning("Unkown class(es) found '{0}' "
```
Use warnings.warn

rragundez committed:
Use logging module to raise warning

jnothman reviewed (commented):
I wrote and lost a longer response. Perhaps don't report the subset ignored. Better each label or no label, because only distinct warning texts will be shown by default.

Now you should probably use a loop over labels with a `try ... except KeyError` rather than use a generator expression. Atm you're duplicating lookups in the mapping.

rragundez committed:
Raise warning for each unkown label

rragundez commented:
@jnothman I've added the changes

jnothman commented on sklearn/preprocessing/label.py:
```diff
@@ -795,7 +797,14 @@ def _transform(self, y, class_mapping):
+                try:
+                    index.add(class_mapping[label])
+                except KeyError:
+                    warnings.warn("Unkown class {0} found and will be ignored"
```
Make that {0!r} to quote the string correctly

jnothman commented on sklearn/preprocessing/tests/test_label.py:
```diff
@@ -307,10 +307,11 @@ def test_multilabel_binarizer_empty_sample():
+    assert_array_equal(mlb.fit(y).transform([[0]]), Y)
 
     mlb = MultiLabelBinarizer(classes=[1, 2])
-    assert_raises(KeyError, mlb.fit_transform, [[0]])
```
Please use assert_warns or a variant

rragundez committed:
Quote string correctly in warning

rragundez committed:
Add unittest for warning if unknown class

rragundez commented:
@jnothman great feedback. Changes done.

jnothman commented on sklearn/preprocessing/tests/test_label.py:
```diff
@@ -307,10 +308,12 @@ def test_multilabel_binarizer_empty_sample():
 
     mlb = MultiLabelBinarizer(classes=[1, 2])
-    assert_raises(KeyError, mlb.fit_transform, [[0]])
+    assert_array_equal(mlb.fit(y).transform([[0]]), Y)
```
We don't want this raising the warning during testing. You can instead get the return value from assert_warns

rragundez replied:
done

jnothman commented on sklearn/preprocessing/label.py:
```diff
@@ -795,7 +797,14 @@ def _transform(self, y, class_mapping):
+                try:
+                    index.add(class_mapping[label])
+                except KeyError:
+                    warnings.warn("Unkown class {0!r} will be ignored"
```
Sorry tofuss over this again, but perhaps it's better still to accumulate all the unknown labels and warn at the end

rragundez replied:
Why? I disagree, there is no internal purpose for the accumulation of these unknown classes. The warning is sufficient for the user to go and double check his/her data.

jnothman reviewed (commented):
Almost there

rragundez committed:
Avoid warnings in tests by using the return value of assert_warns

jnothman commented:
shrug. why? so as to issue only one warning altogether

On 4 Apr 2018 8:56 pm, "rragundez" <notifications@github.com> wrote:

> *@rragundez* commented on this pull request.
> ------------------------------
>
> In sklearn/preprocessing/label.py
> <https://github.com/scikit-learn/scikit-learn/pull/10913#discussion_r179099546>
> :
>
> > @@ -795,7 +797,14 @@ def _transform(self, y, class_mapping):
>          indices = array.array('i')
>          indptr = array.array('i', [0])
>          for labels in y:
> -            indices.extend(set(class_mapping[label] for label in labels))
> +            index = set()
> +            for label in labels:
> +                try:
> +                    index.add(class_mapping[label])
> +                except KeyError:
> +                    warnings.warn("Unkown class {0!r} will be ignored"
>
> Why? I disagree, there is no internal purpose for the accumulation of
> these unknown classes. The warning is sufficient for the user to go and
> double check his/her data.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/10913#discussion_r179099546>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6wJ3L3V5u8WhN2twdFZ4t9o1iiKZks5tlKbQgaJpZM4TFk8K>
> .
>

rragundez commented:
@johannah what benefit that does that have?

rragundez committed:
Accumulate unkown classes and display a single warning

rragundez commented:
@jnothman I made the necessary changes. Do you think is ok now?

jnothman commented:
To not flood the user's screen/log if ignoring is the behaviour they want,
and perhaps to reduce some computational overhead.​

jnothman commented on sklearn/preprocessing/label.py:
```diff
@@ -795,8 +797,17 @@ def _transform(self, y, class_mapping):
         indptr = array.array('i', [0])
         for labels in y:
-            indices.extend(set(class_mapping[label] for label in labels))
+            index, unknown = set(), set()
```
unknown needs to be initialised out of the loop

jnothman commented on sklearn/preprocessing/label.py:
```diff
@@ -795,8 +797,17 @@ def _transform(self, y, class_mapping):
             indptr.append(len(indices))
+        if unknown:
+            warnings.warn("Unkown class(es) {0} will be ignored"
+                          .format(unknown))
```
use sorted for determinism

jnothman commented:
Btw, those last issues could have been caught by tests if you used assert_raises_message

rragundez committed:
Move initialization of unknown classes outside the for loop

rragundez committed:
Sort unknown nclasses on warning output for determinism

In the case that classes are mixed (e.g. 1, 'some_class') sorted
will break if comparing different types, hence the addition of
the key=str argument.

rragundez commented:
@jnothman changes made, and good point about the determinism. BTW I tried the `assert_raises_message` but it seems it works only for Exceptions, is that correct?

rragundez commented:
@jnothman  yeap, I just saw the docstring for `assert_raises_message`: `Helper function to test the message raised in an exception.`

jnothman commented:
I meant assert_warns_message, sorry!

jnothman reviewed (approved):
Otherwise LGTM

rragundez committed:
Modify test to catch warning message for unknown class

rragundez commented:
@jnothman The travis test failed, no idea why. Any update on the merging of this PR?

rragundez committed:
Remove unused import assert_warns

jnothman reviewed (commented):
I think you should also test the case that `classes` is provided. And perhaps your test should ensure that non-ignored labels are still encoded, and that one error message combines those labels ignored from multiple samples.

jnothman commented:
Please add an entry to the change log at `doc/whats_new/v0.20.rst`. Like the other entries there, please reference this pull request with `:issue:` and credit yourself (and other contributors if applicable) with `:user:`

rragundez committed:
Amplify the coverage of the test for several unknown classes

rragundez commented:
@jnothman I did not really understand the request for the test as I think it was already done in here:

https://github.com/rragundez/scikit-learn/blob/b0afb72789f098c5b21b1ec9f683d5d9e2e3798c/sklearn/preprocessing/tests/test_label.py#L317-L320

but I updated the test for multiple unknown classes and the case where `classes` is given.

rragundez committed:
Add method update in sklearn.preprocessing.MultiLabelBinarizer to the change log

jnothman reviewed (approved):
LGTM!

rragundez commented:
@jnothman should we merge it? or are you waiting for another core developer to do that?

qinhanmin2014 reviewed (approved):
LGTM. Thanks @rragundez :)
